File: GFG POTD/020-July-2025/021-count-the-coprimes.java
// https://www.geeksforgeeks.org/problems/count-the-coprimes/1




import java.util.*;

class Solution {
    // TC : O(N + M*log(M))
    // SC : O(M) where M = max(arr[i])
    int cntCoprime(int[] arr) {
        int maxi = Arrays.stream(arr).max().getAsInt();

        int[] freq = new int[maxi + 1];
        for (int num : arr) freq[num] ++;
        
        // count the nos which are divisible by d
        int[] cntDivisibleByD = new int[maxi + 1];
        for (int d = 1; d <= maxi; d ++) {
            for (int mul = d; mul <= maxi; mul += d) {
                cntDivisibleByD[d] += freq[mul];
            }
        }
        
        // dp[d] stores the no of pairs(i, j) such that gcd(arr[i], arr[j]) == 1
        long[] dp = new long[maxi + 1];
        for (int d = maxi; d >= 1; d --) {
            long c = cntDivisibleByD[d];
            dp[d]  = (c * (c - 1)) / 2;   // Total ways to choose 2 elements divisible by d
            
            // Remove overcounted pairs using inclusion-exclusion
            for (int mul = 2*d; mul <= maxi; mul += d) {
                dp[d] -= dp[mul];
            }
        }
        
        // dp[1] contains the number of coprime pairs (gcd = 1)
        return (int)dp[1]; 
    }
}

File: GFG POTD/020-July-2025/021-count-the-coprimes.cpp
// https://www.geeksforgeeks.org/problems/count-the-coprimes/1




#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // TC : O(N + M*log(M))
    // SC : O(M) where M = max(arr[i])
    int cntCoprime(vector<int>& arr) {
        int maxVal = *max_element(begin(arr), end(arr));
        int size = maxVal + 1;

        // Frequency of each number in the array
        vector<int> freq(size, 0);
        for (int num : arr) freq[num]++;

        // Count of elements divisible by each number d
        vector<int> divisibleCount(size, 0);
        for (int d = 1; d < size; d++) {
            for (int multiple = d; multiple < size; multiple += d) {
                divisibleCount[d] += freq[multiple];
            }
        }

        // dp[d] will store number of pairs (i, j) such that gcd(arr[i], arr[j]) = d
        vector<long long> dp(size, 0);
        for (int d = size - 1; d >= 1; d--) {
            long long count = divisibleCount[d];
            dp[d] = count * (count - 1) / 2;  // Total ways to choose 2 elements divisible by d

            // Remove pairs already counted for multiples of d (inclusion-exclusion)
            for (int multiple = 2 * d; multiple < size; multiple += d) {
                dp[d] -= dp[multiple];
            }
        }

        // dp[1] contains number of coprime pairs (gcd = 1)
        return dp[1];
    }
};